Add `autofocus` property to Control

gmikhail opened the pull request:
![autofocus](https://github.com/user-attachments/assets/68a470c8-a99a-4e10-b236-15a6ae9583a9)

- Adding `autofocus` property. The control will attempt to automatically grab focus when it becomes visible. If multiple controls enable `autofocus` property, the last control to become visible will grab focus.
- Adding documentation to `autofocus` property.

Fix https://github.com/godotengine/godot-proposals/issues/6298

---

To test the behavior, you can use the attached project (autofocus is enabled on two buttons, the last one gets focus when the game starts).

[feature-autofocus.zip](https://github.com/user-attachments/files/20532274/feature-autofocus.zip)

gmikhail committed:
Add `autofocus` property to Control

- Adding `autofocus` property. The control will attempt to automatically grab focus after initialization. If multiple controls enable `autofocus` property, the last initialized control will grab focus.
- Adding documentation to `autofocus` property.

Fix https://github.com/godotengine/godot-proposals/issues/6298

gmikhail committed:
Prevent autofocus in editor

Summersay415 commented:
Maybe make it work like `focus_on_show`? It will be more useful (for example, in popup with text field: text field should be focused every time when popup shows, and there is no sense in recreating popup every time to make current realization work).

Summersay415 commented on scene/gui/control.cpp:
```diff
@@ -4083,7 +4108,7 @@ void Control::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "tooltip_auto_translate_mode", PROPERTY_HINT_ENUM, "Inherit,Always,Disabled"), "set_tooltip_auto_translate_mode", "get_tooltip_auto_translate_mode");
 
-	ADD_GROUP("Focus", "focus_");
+	ADD_GROUP("Focus", "");
```
Also, with `focus_on_show` there will be no need in this change

gmikhail commented:
@Summersay415 As discussed [here](https://github.com/godotengine/godot-proposals/issues/6298), most agreed with `autofocus` which is consistent with the HTML attribute name.

Summary of the discussion:
- 4 votes for `autofocus`
- 2 votes for `focus_on_ready`
- One vote for `show_focused`, `focus_on_show`, `grab_focus_on_ready`

Zireael07 commented:
Make that one more vote for autofocus

gmikhail committed:
Improve code style

gmikhail commented:
@AdriaandeJongh The control tries to focus when it receives a NOTIFICATION_POST_ENTER_TREE notification ([documentation](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-constant-notification-post-enter-tree)). English is not my native language, so I may have used "initialization" in the wrong context. I will be glad to see your corrections.

gmikhail committed:
Add more clarification to documentation

Replace the vague "initialization" with the specific "enters the node tree"

gmikhail committed:
Apply autofocus when control becomes visible

not when added to node tree

gmikhail commented:
I added autofocus on visibility change as suggested by @AdriaandeJongh. The accompanying documentation has also been changed.

Updating the header message is in progress...

Summersay415 commented on scene/gui/control.cpp:
```diff
@@ -2724,6 +2724,27 @@ void Control::_window_find_focus_neighbor(const Vector2 &p_dir, Node *p_at, cons
+}
+
+void Control::attempt_autofocus() {
+	if (Engine::get_singleton()->is_editor_hint()) {
+		return;
+	}
```
Should be wrapped in #ifdef TOOLS_ENABLED

gmikhail committed:
Attempt to fix test failure

- Wrap `is_editor_hint()` in `#ifdef TOOLS_ENABLED`
- Documentation element is now in alphabetical order

gmikhail commented:
`🐧 Linux / Editor w/ Mono (target=editor)` now passes successfully.

`🤖 Android / Editor (target=editor) (pull_request)` apparently failed due to an external reason:
```
Downloading https://services.gradle.org/distributions/gradle-8.11.1-bin.zip

Error: Exception in thread "main" java.io.IOException: Downloading from https://services.gradle.org/distributions/gradle-8.11.1-bin.zip failed: timeout
```

bruvzg commented on scene/gui/control.cpp:
```diff
@@ -2724,6 +2724,29 @@ void Control::_window_find_focus_neighbor(const Vector2 &p_dir, Node *p_at, cons
+
+void Control::attempt_autofocus() {
+#ifdef TOOLS_ENABLED
+	if (Engine::get_singleton()->is_editor_hint()) {
```
There's no reason to disallow using it in the editor UI.

```suggestion
	if (is_part_of_edited_scene()) {
```

Calinou replied:
@gmikhail Bump :slightly_smiling_face: Can you look into applying this suggestion?

bruvzg commented on scene/gui/control.cpp:
```diff
@@ -2724,6 +2724,29 @@ void Control::_window_find_focus_neighbor(const Vector2 &p_dir, Node *p_at, cons
+	}
+#endif
+
+	if (get_autofocus() && _is_focus_mode_enabled()) {
```
`_is_focus_mode_enabled` is checking for recursive override only, it's not sufficient to check if control is focusable.

```suggestion
	bool ac_enabled = get_tree() && get_tree()->is_accessibility_enabled();
if (get_autofocus() && ((get_focus_mode_with_override() == FOCUS_ALL) || (get_focus_mode_with_override() == FOCUS_CLICK) || (ac_enabled && get_focus_mode_with_override() == FOCUS_ACCESSIBILITY))) {
```

gmikhail committed:
Improve focus check

gmikhail committed:
Merge branch 'master' into feature-autofocus

KoBeWi commented:
> example usecase: a panel enables a previously hidden sub-panel, so focus needs to move there.

`autofocus` is not a complete solution for this case, assuming you want the 2 panels to be visible together.
You can move focus to the new panel, but once it closes, you need to somehow focus the previous panel. `autofocus` does not provide that.
If both panels are visible, you can navigate focus between the panels. Usually you'd want to avoid it, but changing focus mode involves using code (or some convoluted callback binds), at which point you can focus the new Control manually.
Even if you show only one panel, once the sub-panel is closed, you'd want to go back to the button that opened it, not some fixed button that happen to be autofocus. This again involves code.

So unless I'm missing something, `autofocus` has rather limited usage and realistically it can save a couple lines of code. It does not avoid adding a script, like mentioned in the proposal, because you need code at some point to properly tweak the behavior.

gmikhail committed:
Allow autofocus in editor

Calinou reviewed (commented):
Tested locally, it works as expected. Code looks good to me.

Naming-wise, I'm fine with `autofocus` as it's explicit about what it's doing, but without giving implementation details away (in case we need to change its behavior in the future for some reason). It's also consistent with `autoplay` in Animation and AudioStreamPlayer.

gmikhail committed:
Update doc/classes/Control.xml

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

Calinou reviewed (approved):
Tested again (rebased on top of `master` https://github.com/godotengine/godot/commit/1cf31805370716ce8160ddc1394b9820249a21f7), it works as expected. (I forgot to approve previously, but I confirm this is good on my end.)

Note that following the [mouse-induced visual focus changes in 4.6](https://github.com/godotengine/godot/pull/110250), autofocused controls always get visual focus. This is probably fine, but maybe it's worth a second look from @YeldhamDev just in case.

KoBeWi commented:
Connected joypad does not mean it's being used.

AdriaandeJongh commented:
perhaps this requires a separate proposal, but one way this currently-partial feature could be made complete is if in any autofocus control’s visibility change some system goes over all controls with autofocus, checks whether they are visible, and selects the one rendered on top (taking both hierarchy and z-index into account). but i suspect this is an implementation that’s far away from this current PR…

AThousandShips commented on scene/gui/control.cpp:
```diff
@@ -2724,6 +2724,27 @@ void Control::_window_find_focus_neighbor(const Vector2 &p_dir, Node *p_at, cons
+void Control::set_autofocus(bool p_enable) {
+	if (data.autofocus == p_enable) {
+		return;
+	}
```
```suggestion
```
Unnecessary if the method doesn't do anything with the data like redrawing
